sheepdog/transactions/upload/sub_entities.py

```diff
@@ -190,12 +193,24 @@ def flush_to_session(self):
         # next do node creation
         super(FileUploadEntity, self).flush_to_session()
 
+        # FIXME: we could override rollback() method and revert indexd
+        # changes within it, this would involve some sort of caching and
+        # iterating through all of the affected entities etc. A better
+        # way to go around is to probably implement a dry_run in indexd or
+        # indexclient
         if self.transaction.dry_run:
-            # FIXME: we could override rollback() method and revert indexd
-            # changes within it, this would involve some sort of caching and
-            # iterating through all of the affected entities etc. A better
-            # way to go around is to probably implement a dry_run in indexd or
-            # indexclient
+           return
+
+        # FIXME: In order to avoid incorrect data inside indexd, we skip
+        # further requests to indexd here. However, this won't save us from
+        # the case when something goes wrong during the actual flush: e.g.
+        # out of 20 nodes 10 flushes fine, but then something happens to #11
+        # (networking issue or whatever), then the indexd records for the first
+        # 10 nodes are still going to be changed and we can't rollback them
+        # safely. Implementing some sort of caching might help a bit, but then
+        # again, we will have to send multiple rollback requests to indexd etc,
+        # which may involve retries etc.
+        if not self.transaction.success:
```

Comment:
I don't see an easy fix. This is another case where adding batch update endpoint in indexd would help.

Reply:
Functionally it sounds like you need a queue there to handle that (potential) failure. I understand that that is completely against the whole single transaction...